V gets confused when dealing with Optional Sum Types and spits invalid C code

### Describe the bug

When using `?SumType`, the compiler sometimes gets confused and generates invalid C code. I found two scenarios where this happens.

### Reproduction Steps

1. Casting `?T` to `U` when `T = U | Other`:
```v
type Sum = int | string
fn sum() ?Sum { return Sum(5) }

a := sum()
println(a as int) // V compiler should error, since we're casting from ?Sum to int
```

2. Assigning `T` to `?T` and checking if `?T` is `U` (when `T = U | Other`:
```v
mut a := sum() // a: ?Sum
// ...
a = sum() or { ... } // (sum() or { ... }) converted to ?Sum
if a !is string { ... } // V compiler should error, since we're checking if ?Sum is string
```

### Expected Behavior

The V compiler should error on both situations, preventing the error from propagating into the generated C code, which is harder to analyse

### Current Behavior

In my current codebase, it gave these errors:
```
================== C compilation error (from cc): ==============                                                              cc:        |                                       ~~                                                                            ^~~~~~~~~                                                                                                                  cc: /data/data/com.termux/files/usr/tmp/v_10246/hex.01JGMQ477ZJQZGF038M50E0775.tmp.c:11955:51: error: no member named '_main__Ident' in 'struct _option_main__Token'
cc:  11955 |                 string name = (*(main__Ident*)__as_cast((ident)._main__Ident,(ident)._typ, 111)).ident;
cc:        |                                                         ~~~~~~~ ^
cc: /data/data/com.termux/files/usr/tmp/v_10246/hex.01JGMQ477ZJQZGF038M50E0775.tmp.c:11955:72: error: no member named '_typ' in 'struct _option_main__Token'
cc:  11955 |                 string name = (*(main__Ident*)__as_cast((ident)._main__Ident,(ident)._typ, 111)).ident;
cc:        |                                                                              ~~~~~~~ ^
cc: 3 warnings and 2 errors generated.
... (the original output was 17 lines long, and was truncated to 8 lines)
================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error:
==================
C error found. It should never happen, when compiling pure V code.
```

The code contained both cases in it.

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.9 7b9b3dd

### Environment details (OS name and version, etc.)

V full version: V 0.4.9 b487986.7b9b3dd                         OS: termux, 3.18.140-SIMPLE-KERNEL_V1.1, #1 SMP PREEMPT Sun Apr 18 03:22:59 -03 2021
Processor: 8 cpus, 64bit, little endian                                                                                         getwd: /storage/emulated/0/Documents/hex                        vexe: /data/data/com.termux/files/home/.v/v                     vexe mtime: 2024-12-31 18:09:35                                                                                                 vroot: OK, value: /data/data/com.termux/files/home/.v           VMODULES: OK, value: /data/data/com.termux/files/home/.vmodules
VTMP: OK, value: /data/data/com.termux/files/usr/tmp/v_10246                                                                    Git version: git version 2.47.1                                 Git vroot status: weekly.2024.53 (15 commit(s) behind V master)
.git/config present: true                                       
CC version: clang version 19.1.6                                emcc version: N/A
thirdparty/tcc status: thirdparty-unknown-unknown de82a130
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -2959,10 +2959,17 @@ pub fn (mut c Checker) expr(mut node ast.Expr) ast.Type {
 				c.table.used_features.as_cast = true
 			}
 			if mut node.expr is ast.Ident {
-				if !node.typ.has_flag(.option) && node.expr.obj.typ.has_flag(.option)
-					&& node.expr.or_expr.kind == .absent {
-					c.error('variable `${node.expr.name}` is an Option, it must be unwrapped first',
-						node.expr.pos)
+				if mut node.expr.obj is ast.Var {
+					ident_typ := if node.expr.obj.smartcasts.len > 0 {
+						node.expr.obj.smartcasts.last()
+					} else {
+						node.expr.obj.typ
+					}
+					if !node.typ.has_flag(.option) && ident_typ.has_flag(.option)
+						&& node.expr.or_expr.kind == .absent {
+						c.error('variable `${node.expr.name}` is an Option, it must be unwrapped first',
+							node.expr.pos)
+					}
 				}
 			}
 			if expr_type_sym.kind == .sum_type {

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -1386,6 +1386,13 @@ fn (mut c Checker) check_expr_option_or_result_call(expr ast.Expr, ret_type ast.
 			c.check_expr_option_or_result_call(expr.expr, ret_type)
 		}
 		ast.AsCast {
+			if expr.expr is ast.Ident {
+				if !ret_type.has_flag(.option) && expr.expr.obj.typ.has_flag(.option)
+					&& expr.expr.or_expr.kind == .absent {
+					c.error('variable `${expr.expr.name}` is an Option, it must be unwrapped first',
+						expr.expr.pos)
+				}
+			}
 			c.check_expr_option_or_result_call(expr.expr, ret_type)
 		}
 		ast.ParExpr {
@@ -2958,6 +2965,7 @@ pub fn (mut c Checker) expr(mut node ast.Expr) ast.Type {
 			if !c.is_builtin_mod {
 				c.table.used_features.as_cast = true
 			}
+			c.check_expr_option_or_result_call(node.expr, node.typ)
 			if expr_type_sym.kind == .sum_type {
 				c.ensure_type_exists(node.typ, node.pos)
 				if !c.table.sumtype_has_variant(c.unwrap_generic(node.expr_type), c.unwrap_generic(node.typ),

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -1386,13 +1386,6 @@ fn (mut c Checker) check_expr_option_or_result_call(expr ast.Expr, ret_type ast.
 			c.check_expr_option_or_result_call(expr.expr, ret_type)
 		}
 		ast.AsCast {
-			if expr.expr is ast.Ident {
-				if !ret_type.has_flag(.option) && expr.expr.obj.typ.has_flag(.option)
-					&& expr.expr.or_expr.kind == .absent {
-					c.error('variable `${expr.expr.name}` is an Option, it must be unwrapped first',
-						expr.expr.pos)
-				}
-			}
 			c.check_expr_option_or_result_call(expr.expr, ret_type)
 		}
 		ast.ParExpr {
@@ -2965,7 +2958,13 @@ pub fn (mut c Checker) expr(mut node ast.Expr) ast.Type {
 			if !c.is_builtin_mod {
 				c.table.used_features.as_cast = true
 			}
-			c.check_expr_option_or_result_call(node.expr, node.typ)
+			if mut node.expr is ast.Ident {
+				if !node.typ.has_flag(.option) && node.expr.obj.typ.has_flag(.option)
+					&& node.expr.or_expr.kind == .absent {
+					c.error('variable `${node.expr.name}` is an Option, it must be unwrapped first',
+						node.expr.pos)
+				}
+			}
 			if expr_type_sym.kind == .sum_type {
 				c.ensure_type_exists(node.typ, node.pos)
 				if !c.table.sumtype_has_variant(c.unwrap_generic(node.expr_type), c.unwrap_generic(node.typ),

--- a/vlib/v/checker/tests/option_as_cast_err.out
+++ b/vlib/v/checker/tests/option_as_cast_err.out
@@ -1,0 +1,5 @@
+vlib/v/checker/tests/option_as_cast_err.vv:12:9: error: variable `b` is an Option, it must be unwrapped first
+   10 | println(b)
+   11 | println(b? as int)
+   12 | println(b as int)
+      |         ^

--- a/vlib/v/checker/tests/option_as_cast_err.vv
+++ b/vlib/v/checker/tests/option_as_cast_err.vv
@@ -1,0 +1,12 @@
+type Sum = int | string
+
+fn sum() ?Sum {
+	return Sum(5)
+}
+
+a := sum()?
+println(a as int)
+b := sum()
+println(b)
+println(b? as int)
+println(b as int)

--- a/vlib/v/tests/options/option_concatexpr_test.v
+++ b/vlib/v/tests/options/option_concatexpr_test.v
@@ -12,17 +12,17 @@ fn foo(f int) !(int, ?int) {
 
 fn test_main() {
 	a, b := foo(-1) or { 2, 2 }
-	c := b as int
+	c := b? as int
 	assert a == 2
 	assert c == 2
 
 	a2, b2 := foo(0) or { 2, 2 }
-	c2 := b2 as int
+	c2 := b2? as int
 	assert a2 == 0
 	assert c2 == 0
 
 	a3, b3 := foo(1) or { 2, 2 }
-	c3 := b3 as int
+	c3 := b3? as int
 	assert a3 == 1
 	assert c3 == 1
 }

--- a/vlib/v/tests/options/option_unwrap_as_cast_test.v
+++ b/vlib/v/tests/options/option_unwrap_as_cast_test.v
@@ -15,9 +15,9 @@ fn get_opt_int(a int) ?int {
 fn test_option_unwrap_as_cast() {
 	x := get_opt(1)
 	d := get_opt_int(12)
-	dump(d as int == 12)
-	dump('${x as string}' == 'success')
+	dump(d? as int == 12)
+	dump('${x? as string}' == 'success')
 
-	assert d as int == 12
-	assert x as string == 'success'
+	assert d? as int == 12
+	assert x? as string == 'success'
 }

